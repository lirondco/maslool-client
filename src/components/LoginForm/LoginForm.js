import React, { Component } from 'react'
import { Input, Label, Button } from '../Utils/Utils'
import AuthApiService from '../../services/auth-api-service'
import UserContext from '../../contexts/UserContext'
import './LoginForm.css'
import Loading from '../Loading/Loading'

export default class LoginForm extends Component {
    static defaultProps = {
        onLoginSuccess: () => { }
    }

    static contextType = UserContext

    state = {
        error: null,
        isLodaing: false
    }

    firstInput = React.createRef()

    handleSubmit = ev => {
        ev.preventDefault()
        const { username, password } = ev.target

        this.setState({ error: "Processing your request ..." })

        AuthApiService.postLogin({
            username: username.value,
            password: password.value
        })
            .then(res => {
                this.setState({ error: null, isLoading: true })
                username.value = ''
                password.value = ''
                this.context.processLogin(res.authToken)
                this.props.onLoginSuccess()
            })
            .then(this.setState({ isLoading: false }))
            .catch(res => {
                this.setState({ error: res.error })
            })
    }

    componentDidMount() {
        this.firstInput.current.focus()
    }

    render() {
        const { error } = this.state
        if (this.state.isLoading === true) {
            return <Loading />
        }
        return (
            <>
                <h2>Sign in</h2>
                <hr />
                <form 
                    className='LoginForm'
                    onSubmit={this.handleSubmit}
                >
                    <div>
                        <Label htmlFor='login-username-input'>
                            Username: 
                        </Label>
                        <Input 
                            ref={this.firstInput}
                            id='login-username-input'
                            name='username'
                            required
                        />
                        <Label htmlFor='login-password-input'>
                            Password:
                        </Label>
                        <Input 
                            id='login-password-input'
                            name='password'
                            type='password'
                            required
                        />
                    </div>
                    <div role='alert'>
                        {error && <p className='error'>{error}</p>}
                    </div>
                    <Button type='submit'>
                        Sign in
                    </Button>
                </form>
            </>
        )
    }
}