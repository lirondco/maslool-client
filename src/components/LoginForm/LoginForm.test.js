import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import LoginForm from "./LoginForm";
import { withContext } from "shallow-with-context";

describe(`LoginForm component`, () => {

  const context = {
    clearError: () => {},
    user: { id: 3, admin: false, banned: false, username: "test_user" },
  };
  const ComponentWithContext = withContext(LoginForm, context);

  it("renders the form", () => {
    const wrapper = shallow(<ComponentWithContext />, { context } );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("form submits when submitted", () => {
    const wrapper = shallow(<ComponentWithContext />, { context });
    const mockedEvent = {
      preventDefault: () => {},
      target: { username: { value: "Test" }, password: { value: "test" } },
    };
    wrapper.find("form").at(0).simulate("submit", mockedEvent);
  });
});
