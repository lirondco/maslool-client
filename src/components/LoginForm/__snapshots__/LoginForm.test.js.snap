// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm component renders the form 1`] = `
<Fragment>
  <h2>
    Sign in
  </h2>
  <hr />
  <form
    className="LoginForm"
    onSubmit={[Function]}
  >
    <div>
      <Label
        htmlFor="login-username-input"
      >
        Username:
      </Label>
      <ForwardRef
        id="login-username-input"
        name="username"
        required={true}
      />
      <Label
        htmlFor="login-password-input"
      >
        Password:
      </Label>
      <ForwardRef
        id="login-password-input"
        name="password"
        required={true}
        type="password"
      />
    </div>
    <div
      role="alert"
    />
    <ForwardRef
      type="submit"
    >
      Sign in
    </ForwardRef>
  </form>
</Fragment>
`;
